crates/resolver/src/name_server/name_server_pool.rs

```diff
@@ -215,18 +164,21 @@ async fn try_send<P: ConnectionProvider>(
     // TODO: more principled handling of timeouts. Currently, timeouts appear to be handled mostly
     // close to the connection, which means the top level resolution might take substantially longer
     // to fire than the timeout configured in `ResolverOpts`.
+    let mut conns = VecDeque::from(conns);
     let mut backoff = Duration::from_millis(20);
     let mut busy = SmallVec::<[NameServer<P>; 2]>::new();
     let mut err = ProtoError::from(ProtoErrorKind::NoConnections);
+    let mut skip_udp = false;
 
     loop {
         // construct the parallel requests, 2 is the default
         let mut par_conns = SmallVec::<[NameServer<P>; 2]>::new();
-        let count = conns.len().min(opts.num_concurrent_reqs.max(1));
-
-        // Shuffe DNS NameServers to avoid overloads to the first configured ones
-        for conn in conns.drain(..count) {
-            par_conns.push(conn);
+        while !conns.is_empty() && par_conns.len() < Ord::max(opts.num_concurrent_reqs, 1) {
+            if let Some(conn) = conns.pop_front() {
+                if !(skip_udp && conn.protocol() == Protocol::Udp) {
+                    par_conns.push(conn);
+                }
```

Comment:
One other thought about preferring UDP over TCP: if `num_concurrent_reqs` is greater than the number of UDP connections, this will start by making concurrent UDP and TCP connections. Those could further end up being two concurrent connections to the same resolver IP, which would be wasteful. I think this might address it.
```suggestion
        while !conns.is_empty() && par_conns.len() < Ord::max(opts.num_concurrent_reqs, 1) {
            if let Some(conn) = conns.front() {
                if !skip_udp && !par_conns.is_empty() && conn.protocol() =! Protocol::Udp {
                    // If making UDP queries, and there are no more UDP connections to add, stop.
                    break;
                }
            }
            if let Some(conn) = conns.pop_front() {
                if !(skip_udp && conn.protocol() == Protocol::Udp) {
                    par_conns.push(conn);
                }
```

Reply:
OK, addressing this at another level sounds good